src/coreclr/vm/jithelpers.cpp

```diff
@@ -1793,6 +1804,77 @@ HCIMPL2(void*, JIT_GetSharedNonGCThreadStaticBase, DomainLocalModule *pDomainLoc
     return HCCALL1(JIT_GetNonGCThreadStaticBase_Helper, pMT);
 }
 HCIMPLEND
+
+// *** This helper corresponds CORINFO_HELP_GETSHARED_NONGCTHREADSTATIC_BASE_NOCTOR_OPTIMIZED.
+//      Even though we always check if the class constructor has been run, we have a separate
+//      helper ID for the "no ctor" version because it allows the JIT to do some reordering that
+//      otherwise wouldn't be possible.
+HCIMPL3(void*, JIT_GetSharedNonGCThreadStaticBaseOptimized, DomainLocalModule *pDomainLocalModule, DWORD dwClassDomainID, UINT32 staticBlockIndex)
```

Comment:
This helper only really needs `staticBlockIndex`. The remaining arguments are redundant - they can be computed from staticBlockIndex.

Reply:
Are you suggesting something like this?

```cpp
MethodTable * pMT = CEEInfo::g_threadStaticBlockTypeIDMap.LookupType(staticBlockIndex);
staticBlock = HCCALL1(JIT_GetNonGCThreadStaticBase_Helper, pMT);
```

Reply:
Yes, something like this.

Reply:
Done